File: 307.cpp
// Given an integer array nums, find the sum of the elements between indices i
// and j (i ≤ j), inclusive.
//
// The update(i, val) function modifies nums by updating the element at index i
// to val.
//
// Example:
//
// Given nums = [1, 3, 5]
//
// sumRange(0, 2) -> 9
// update(1, 2)
// sumRange(0, 2) -> 8
// Note:
//
// The array is only modifiable by the update function.
// You may assume the number of calls to update and sumRange function is
// distributed evenly.

class NumArray {
public:
    struct SegmentTreeNode {
        int start;
        int end;
        int sum;
        SegmentTreeNode* left;
        SegmentTreeNode* right;
        SegmentTreeNode(int s, int e, int sm = 0)
          : start(s), end(e), sum(sm), left(nullptr), right(nullptr) {}
    };

    SegmentTreeNode* root;

    NumArray(vector<int> nums) {
        int n = nums.size();
        root = buildTree(nums, 0, n - 1);
    }

    SegmentTreeNode* buildTree(vector<int>& nums, int start, int end) {
        if (start > end) {
            return nullptr;
        }

        SegmentTreeNode* root = new SegmentTreeNode(start, end);
        if (start == end) {
            root->sum = nums[start];
            return root;
        }

        int mid = start + (end - start) / 2;
        root->left = buildTree(nums, start, mid);
        root->right = buildTree(nums, mid + 1, end);
        root->sum = root->left->sum + root->right->sum;
        return root;
    }

    int updateTree(int i, int val, SegmentTreeNode* root) {
        if (root == nullptr) {
            return 0;
        }

        int diff;
        if (root->start == i && root->end == i) {
            diff = val - root->sum;
            root->sum = val;
            return diff;
        }

        int mid = root->start + (root->end - root->start) / 2;
        if (i <= mid) {
            diff = updateTree(i, val, root->left);
        } else {
            diff = updateTree(i, val, root->right);
        }
        root->sum += diff;
        return diff;
    }

    void update(int i, int val) { updateTree(i, val, root); }

    int sumRangeTree(int i, int j, SegmentTreeNode* root) {
        if (root == nullptr) {
            return 0;
        }

        if (i > root->end || j < root->start) {
            return 0;
        }

        if (root->start == i && root->end == j) {
            return root->sum;
        }

        int mid = root->start + (root->end - root->start) / 2;
        if (i > mid) {
            return sumRangeTree(i, j, root->right);
        }
        if (j <= mid) {
            return sumRangeTree(i, j, root->left);
        }

        return sumRangeTree(i, mid, root->left)
               + sumRangeTree(mid + 1, j, root->right);
    }

    int sumRange(int i, int j) { return sumRangeTree(i, j, root); }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * obj.update(i,val);
 * int param_2 = obj.sumRange(i,j);
 */

File: Python3/307.py
class NumArray:
    class Node:
        def __init__(self, start, end, val, left=None, right=None):
            self.start = start
            self.end = end
            self.val = val
            self.left = left
            self.right = right

    def __init__(self, nums: List[int]):
        self.root = self._buildSegmentTree(nums, 0, len(nums)-1)
    
    def _buildSegmentTree(self, nums, i, j):
        if i>j:
            return None
        if i == j:
            return self.Node(i, j, nums[i])
        mid = (i+j)//2
        left = self._buildSegmentTree(nums, i, mid)
        right = self._buildSegmentTree(nums, mid+1, j)
        val = 0
        if left:
            val += left.val
        if right:
            val += right.val
        return self.Node(i, j, val, left, right)
        

    def update(self, index: int, val: int) -> None:
        self._updateSegmentTree(self.root, index, val)
    
    def _updateSegmentTree(self, root, index, val):
        if not root:
            return 0
        if root.start == index and root.end == index:
            diff = val - root.val
            root.val = val
            return diff
        mid = (root.start+root.end)//2
        diff = 0
        if index <= mid:
            diff = self._updateSegmentTree(root.left, index, val)
        else:
            diff = self._updateSegmentTree(root.right, index, val)
        root.val += diff
        return diff
        

    def sumRange(self, left: int, right: int) -> int:
        return self._find(self.root, left, right)

    def _find(self, root, i, j):
        if root.start == i and root.end == j:
            return root.val
        mid = (root.start+root.end)//2
        if j<=mid:
            return self._find(root.left, i, j)
        if i>mid:
            return self._find(root.right, i, j)
        return self._find(root.left, i, mid) + self._find(root.right, mid+1, j)



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)
